src/main/java/org/testng/internal/ExceptionUtils.java

```diff
@@ -0,0 +1,58 @@
+package org.testng.internal;
+
+import org.testng.IInvokedMethod;
+import org.testng.ITestContext;
+import org.testng.ITestNGMethod;
+import org.testng.ITestResult;
+
+import java.util.Set;
+
+/**
+ * A utility class that houses mechanisms to retrieve exception information.
+ */
+final class ExceptionUtils {
+    public ExceptionUtils() {
+        //Utility class. Defeat instantiation.
+    }
+
+    static Throwable getExceptionDetails(ITestContext context, Object instance) {
+        Set<ITestResult> configResults = context.getFailedConfigurations().getAllResults();
+        if (configResults.isEmpty()) {
+            configResults = context.getSkippedConfigurations().getAllResults();
```

Comment:
I am just thinking about the scenario wherein there is both failure and skip on the same instance. Since we are working with a Set or order is perhaps not guaranteed with order, we may end up retrieving the skipped exception details instead of the failed exception details.

Reply:
Ok, but the logic is not easy to understand.
What if we have mixed failed and skipped?
About order, you may use a temp list, or just add some comments

Reply:
The current implementation will always give priority to failed configurations and will fall back to skipped ones only when there are no failed configurations. So if there are a combination of failed and skipped, failed ones are what get considered because they are the ones that need to be fixed before a user can move forward towards addressing the skipped ones.